module/icp/include/modes/modes.h

```diff
@@ -244,6 +291,54 @@ typedef struct gcm_ctx {
 #define	AES_GMAC_IV_LEN		12
 #define	AES_GMAC_TAG_BITS	128
 
+/*
+ * Clear sensitive data in the context and free allocated memory.
+ *
+ * ctx->gcm_remainder may contain a plaintext remainder. ctx->gcm_H and
+ * ctx->gcm_Htable contain the hash sub key which protects authentication.
+ * ctx->gcm_pt_buf contains the plaintext result of decryption.
+ *
+ * Although extremely unlikely, ctx->gcm_J0 and ctx->gcm_tmp could be used for
+ * a known plaintext attack, they consists of the IV and the first and last
+ * counter respectively. If they should be cleared is debatable.
+ */
+static inline void
+gcm_clear_ctx(gcm_ctx_t *ctx)
+{
+	memset(ctx->gcm_remainder, 0, sizeof (ctx->gcm_remainder));
+	memset(ctx->gcm_H, 0, sizeof (ctx->gcm_H));
+#if defined(CAN_USE_GCM_ASM)
+	if (ctx->gcm_simd_impl != GSI_NONE) {
+		ASSERT3P(ctx->gcm_Htable, !=, NULL);
+		memset(ctx->gcm_Htable, 0, ctx->gcm_htab_len);
+		kmem_free(ctx->gcm_Htable, ctx->gcm_htab_len);
+	}
+#endif
+	if (ctx->gcm_pt_buf != NULL) {
+		memset(ctx->gcm_pt_buf, 0, ctx->gcm_pt_buf_len);
+		vmem_free(ctx->gcm_pt_buf, ctx->gcm_pt_buf_len);
+	}
+	/* Optional */
+	memset(ctx->gcm_J0, 0, sizeof (ctx->gcm_J0));
+	memset(ctx->gcm_tmp, 0, sizeof (ctx->gcm_tmp));
+
+#ifdef DEBUG_GCM_ASM
+	if (ctx->gcm_shadow_ctx != NULL) {
+		/* No need to clear data while debugging, just free memory. */
+		gcm_ctx_t *sc = ctx->gcm_shadow_ctx;
+
+		if (sc->gcm_Htable != NULL) {
+			kmem_free(c->gcm_Htable, c->gcm_htab_len);
+		}
+		if (sc->gcm_pt_buf != NULL) {
+			vmem_free(c->gcm_pt_buf, c->gcm_pt_buf_len);
+		}
```

Comment:
Should it be `sc->` instead of `c->` in the free arguments?

Reply:
Yes, nice catch. I've this fix in my tree already, dunno why I missed it here.

Reply:
Looks like a merge error. Fixed.